core/trino-main/src/main/java/io/trino/sql/planner/optimizations/SkewJoinOptimizer.java

```diff
@@ -0,0 +1,440 @@
+
+        public SkewedContext(List<String> values, DataType keyType)
+        {
+            this.keyType = keyType;
+            this.values = values;
+        }
+    }
+
+    static class PlanRewriter
+            extends PlanVisitor<PlanNode, Map<Symbol, SkewedContext>>
+    {
+        private final Session session;
+        private final Metadata metadata;
+        private final JoinRewriter joinRewriter;
+        private final Map<String, SkewJoinConfig> skewedTables;
+
+        public PlanRewriter(Session session, Metadata metadata, Map<String, SkewJoinConfig> skewedTables, JoinRewriter rewriter)
+        {
+            this.session = session;
+            this.metadata = metadata;
+            this.skewedTables = skewedTables;
+            this.joinRewriter = rewriter;
+        }
+
+        public PlanNode passThrough(PlanNode node, Map<Symbol, SkewedContext> context)
+        {
+            List<PlanNode> rewrittenChildren = node.getSources().stream()
+                    .map(child -> child.accept(this, context))
+                    .collect(Collectors.toList());
+            return replaceChildren(node, rewrittenChildren);
+        }
+
+        @Override
+        public PlanNode visitJoin(JoinNode node, Map<Symbol, SkewedContext> context)
+        {
+            // TODO : how do range joins behave ?
+            Map<Symbol, SkewedContext> leftCtx = new HashMap<>();
+            Map<Symbol, SkewedContext> rightCtx = new HashMap<>();
+
+            PlanNode leftReWritten = node.getLeft().accept(this, leftCtx);
+            PlanNode rightReWritten = node.getRight().accept(this, rightCtx);
+
+            JoinNode newJoin = this.joinRewriter.rewriteJoin(node, leftReWritten, rightReWritten, Stream.empty());
+
+            // TODO: Currently we only handle the case when the skewed symbol is on the probe side. lets also handle the stream side.
+            if (leftCtx.size() > 0) {
+                Optional<JoinNode.EquiJoinClause> maybeClause = newJoin.getCriteria().stream()
+                        .filter(clause -> leftCtx.containsKey(clause.getLeft()))
+                        .findFirst();
+
+                if (maybeClause.isPresent()) {
+                    JoinNode.EquiJoinClause skewedClause = maybeClause.get();
+                    return this.joinRewriter.rewriteSkewedJoin(newJoin, skewedClause, leftCtx.get(skewedClause.getLeft()));
+                }
+            }
+
+            return newJoin;
+        }
+
+        @Override
+        public PlanNode visitFilter(FilterNode node, Map<Symbol, SkewedContext> context)
+        {
+            return passThrough(node, context);
+        }
+
+        @Override
+        public PlanNode visitSort(SortNode node, Map<Symbol, SkewedContext> context)
+        {
+            return passThrough(node, context);
+        }
+
+        @Override
+        public PlanNode visitExchange(ExchangeNode node, Map<Symbol, SkewedContext> context)
+        {
+            return passThrough(node, context);
```

Comment:
This is not correct. ExchangeNode remaps input symbols to output symbols.

However, since this optimizer needs to run before exchanges are added. I'd remove this and just fail with an error in case this node is encountered.

Reply:
good point, will fail here.